Update CIFAR10 example

nickgnd opened the pull request:
Hey all 👋 

I wanted to experiment a bit with the CIFAR10 dataset using as starting point the example in the repo, but when I ran it I got back the following error:

```
❯ elixir examples/vision/cifar10.exs

00:26:35.804 [info] XLA service 0x7f94db905030 initialized for platform Host (this does not guarantee that XLA will be used). Devices:

00:26:35.806 [info]   StreamExecutor device (0): Host, Default Version
** (FunctionClauseError) no function clause matching in Axon.input/2

    The following arguments were given to Axon.input/2:

        # 1
        "input"

        # 2
        [shape: {nil, 3, 32, 32}]

    Attempted function clauses (showing 1 out of 1):

        def input(input_shape, name) when is_tuple(input_shape) and is_binary(name)

    (axon 0.1.0) lib/axon.ex:411: Axon.input/2
    examples/vision/cifar10.exs:33: Cifar.build_model/1
    examples/vision/cifar10.exs:66: Cifar.run/0
```

Therefore, I tried to update the implementation to use the latest latest version of Nx and Axon.

While updating the deps, I also update the implementation to:
* use `Nx.to_batched` instead of the deprecated `Nx.to_batched_list`
* transpose the inputs to have the `channels` axis as last to don't change the model implementation (note: Axon Conv layer [expects channels as last by default](https://github.com/elixir-nx/axon/blob/7c21ae75d34e68c946a8ff24a46f194f9d17003a/lib/axon/layers.ex#L356)).

To conclude, below you can find the output of the training and testing:

```
❯ elixir examples/vision/cifar10.exs

12:06:13.711 [info] TfrtCpuClient created.
#Axon<
  inputs: %{"input" => {nil, 32, 32, 3}}
  outputs: "softmax_0"
  nodes: 15
>


 Training Model


12:06:14.758 [debug] Forwarding options: [compiler: EXLA] to JIT compiler
Epoch: 0, Batch: 1450, Accuracy: 0.3524725 loss: 1.7826751
Epoch: 1, Batch: 1450, Accuracy: 0.4720236 loss: 1.6165651
Epoch: 2, Batch: 1450, Accuracy: 0.5339851 loss: 1.5129054
Epoch: 3, Batch: 1450, Accuracy: 0.5710936 loss: 1.4376966
Epoch: 4, Batch: 1450, Accuracy: 0.5962912 loss: 1.3778347
Epoch: 5, Batch: 1450, Accuracy: 0.6215758 loss: 1.3282862
Epoch: 6, Batch: 1450, Accuracy: 0.6390857 loss: 1.2851359
Epoch: 7, Batch: 1450, Accuracy: 0.6504776 loss: 1.2472833
Epoch: 8, Batch: 1450, Accuracy: 0.6667596 loss: 1.2139190
Epoch: 9, Batch: 1450, Accuracy: 0.6782400 loss: 1.1829219


 Testing Model


12:17:00.656 [debug] Forwarding options: [compiler: EXLA] to JIT compiler
Batch: 62, Accuracy: 0.6825397
```

Let me know if there is anything else I can do.
As always, thanks again for all the great work you are doing! 🙇 

Best.

seanmor5 reviewed (commented):
It may be best to actually change the input pipeline to use images with channels last configuration rather than change the model

nickgnd commented:
Hey @seanmor5 thanks for feedback!

I updated the PR and transposed the inputs to have the `channels` axis as last in order to keep the model as it was before. I also updated the PR description accordingly. Please, let me know if this was your idea or if you had something else in mind, thanks 🙇 

Have a great day! ✌️ 

---

Btw, I was too naive, but I didn't expect such big difference accuracy-wise between the 2 models (channels first --> Accuracy: 0.6825397  Vs.  channels last --> Accuracy: 0.0927579) 💡 Actually I expected similar results, is it a best practice to keep channels as last axis?

I printed the models as table (see below) to get more insights, and I can see differences, but I can not finger point where this accuracy gap is originated.

<details>
  <summary>Model recap when channels `:last`</summary>

```
+---------------------------------------------------------------------------------------------------------------------------------------+
|                                                                 Model                                                                 |
+=========================================+====================+==================+=========================+===========================+
| Layer                                   | Input Shape        | Output Shape     | Options                 | Parameters                |
+=========================================+====================+==================+=========================+===========================+
| input ( input )                         | []                 | {32, 32, 32, 3}  | shape: {nil, 32, 32, 3} |                           |
|                                         |                    |                  | optional: false         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| conv_0 ( conv["input"] )                | [{32, 32, 32, 3}]  | {32, 30, 30, 32} | strides: 1              | kernel: f32[3][3][3][32]  |
|                                         |                    |                  | padding: :valid         | bias: f32[32]             |
|                                         |                    |                  | input_dilation: 1       |                           |
|                                         |                    |                  | kernel_dilation: 1      |                           |
|                                         |                    |                  | feature_group_size: 1   |                           |
|                                         |                    |                  | channels: :last         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| relu_0 ( relu["conv_0"] )               | [{32, 30, 30, 32}] | {32, 30, 30, 32} |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| batch_norm_0 ( batch_norm["relu_0"] )   | [{32, 30, 30, 32}] | {32, 30, 30, 32} | epsilon: 1.0e-5         | gamma: f32[32]            |
|                                         |                    |                  | channel_index: -1       | beta: f32[32]             |
|                                         |                    |                  | momentum: 0.1           | mean: f32[32]             |
|                                         |                    |                  |                         | var: f32[32]              |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| max_pool_0 ( max_pool["batch_norm_0"] ) | [{32, 30, 30, 32}] | {32, 15, 15, 32} | kernel_size: {2, 2}     |                           |
|                                         |                    |                  | strides: nil            |                           |
|                                         |                    |                  | padding: :valid         |                           |
|                                         |                    |                  | channels: :last         |                           |
|                                         |                    |                  | window_dilations: 1     |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| conv_1 ( conv["max_pool_0"] )           | [{32, 15, 15, 32}] | {32, 13, 13, 64} | strides: 1              | kernel: f32[3][3][32][64] |
|                                         |                    |                  | padding: :valid         | bias: f32[64]             |
|                                         |                    |                  | input_dilation: 1       |                           |
|                                         |                    |                  | kernel_dilation: 1      |                           |
|                                         |                    |                  | feature_group_size: 1   |                           |
|                                         |                    |                  | channels: :last         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| relu_1 ( relu["conv_1"] )               | [{32, 13, 13, 64}] | {32, 13, 13, 64} |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| batch_norm_1 ( batch_norm["relu_1"] )   | [{32, 13, 13, 64}] | {32, 13, 13, 64} | epsilon: 1.0e-5         | gamma: f32[64]            |
|                                         |                    |                  | channel_index: -1       | beta: f32[64]             |
|                                         |                    |                  | momentum: 0.1           | mean: f32[64]             |
|                                         |                    |                  |                         | var: f32[64]              |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| max_pool_1 ( max_pool["batch_norm_1"] ) | [{32, 13, 13, 64}] | {32, 6, 6, 64}   | kernel_size: {2, 2}     |                           |
|                                         |                    |                  | strides: nil            |                           |
|                                         |                    |                  | padding: :valid         |                           |
|                                         |                    |                  | channels: :last         |                           |
|                                         |                    |                  | window_dilations: 1     |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| flatten_0 ( flatten["max_pool_1"] )     | [{32, 6, 6, 64}]   | {32, 2304}       |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| dense_0 ( dense["flatten_0"] )          | [{32, 2304}]       | {32, 64}         |                         | kernel: f32[2304][64]     |
|                                         |                    |                  |                         | bias: f32[64]             |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| relu_2 ( relu["dense_0"] )              | [{32, 64}]         | {32, 64}         |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| dropout_0 ( dropout["relu_2"] )         | [{32, 64}]         | {32, 64}         | rate: 0.5               | key: f32[2]               |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| dense_1 ( dense["dropout_0"] )          | [{32, 64}]         | {32, 10}         |                         | kernel: f32[64][10]       |
|                                         |                    |                  |                         | bias: f32[10]             |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| softmax_0 ( softmax["dense_1"] )        | [{32, 10}]         | {32, 10}         |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
Total Parameters: 167948
Total Parameters Memory: 671792 bytes
```
</details>

<details>
  <summary>Model recap when channels `:first`</summary>

```
+---------------------------------------------------------------------------------------------------------------------------------------+
|                                                                 Model                                                                 |
+=========================================+====================+==================+=========================+===========================+
| Layer                                   | Input Shape        | Output Shape     | Options                 | Parameters                |
+=========================================+====================+==================+=========================+===========================+
| input ( input )                         | []                 | {32, 3, 32, 32}  | shape: {nil, 3, 32, 32} |                           |
|                                         |                    |                  | optional: false         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| conv_0 ( conv["input"] )                | [{32, 3, 32, 32}]  | {32, 32, 30, 30} | strides: 1              | kernel: f32[32][3][3][3]  |
|                                         |                    |                  | padding: :valid         | bias: f32[32]             |
|                                         |                    |                  | input_dilation: 1       |                           |
|                                         |                    |                  | kernel_dilation: 1      |                           |
|                                         |                    |                  | feature_group_size: 1   |                           |
|                                         |                    |                  | channels: :first        |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| relu_0 ( relu["conv_0"] )               | [{32, 32, 30, 30}] | {32, 32, 30, 30} |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| batch_norm_0 ( batch_norm["relu_0"] )   | [{32, 32, 30, 30}] | {32, 32, 30, 30} | epsilon: 1.0e-5         | gamma: f32[30]            |
|                                         |                    |                  | channel_index: -1       | beta: f32[30]             |
|                                         |                    |                  | momentum: 0.1           | mean: f32[30]             |
|                                         |                    |                  |                         | var: f32[30]              |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| max_pool_0 ( max_pool["batch_norm_0"] ) | [{32, 32, 30, 30}] | {32, 16, 15, 30} | kernel_size: {2, 2}     |                           |
|                                         |                    |                  | strides: nil            |                           |
|                                         |                    |                  | padding: :valid         |                           |
|                                         |                    |                  | channels: :last         |                           |
|                                         |                    |                  | window_dilations: 1     |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| conv_1 ( conv["max_pool_0"] )           | [{32, 16, 15, 30}] | {32, 64, 13, 28} | strides: 1              | kernel: f32[64][16][3][3] |
|                                         |                    |                  | padding: :valid         | bias: f32[64]             |
|                                         |                    |                  | input_dilation: 1       |                           |
|                                         |                    |                  | kernel_dilation: 1      |                           |
|                                         |                    |                  | feature_group_size: 1   |                           |
|                                         |                    |                  | channels: :first        |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| relu_1 ( relu["conv_1"] )               | [{32, 64, 13, 28}] | {32, 64, 13, 28} |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| batch_norm_1 ( batch_norm["relu_1"] )   | [{32, 64, 13, 28}] | {32, 64, 13, 28} | epsilon: 1.0e-5         | gamma: f32[28]            |
|                                         |                    |                  | channel_index: -1       | beta: f32[28]             |
|                                         |                    |                  | momentum: 0.1           | mean: f32[28]             |
|                                         |                    |                  |                         | var: f32[28]              |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| max_pool_1 ( max_pool["batch_norm_1"] ) | [{32, 64, 13, 28}] | {32, 32, 6, 28}  | kernel_size: {2, 2}     |                           |
|                                         |                    |                  | strides: nil            |                           |
|                                         |                    |                  | padding: :valid         |                           |
|                                         |                    |                  | channels: :last         |                           |
|                                         |                    |                  | window_dilations: 1     |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| flatten_0 ( flatten["max_pool_1"] )     | [{32, 32, 6, 28}]  | {32, 5376}       |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| dense_0 ( dense["flatten_0"] )          | [{32, 5376}]       | {32, 64}         |                         | kernel: f32[5376][64]     |
|                                         |                    |                  |                         | bias: f32[64]             |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| relu_2 ( relu["dense_0"] )              | [{32, 64}]         | {32, 64}         |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| dropout_0 ( dropout["relu_2"] )         | [{32, 64}]         | {32, 64}         | rate: 0.5               | key: f32[2]               |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| dense_1 ( dense["dropout_0"] )          | [{32, 64}]         | {32, 10}         |                         | kernel: f32[64][10]       |
|                                         |                    |                  |                         | bias: f32[10]             |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
| softmax_0 ( softmax["dense_1"] )        | [{32, 10}]         | {32, 10}         |                         |                           |
+-----------------------------------------+--------------------+------------------+-------------------------+---------------------------+
Total Parameters: 355188
Total Parameters Memory: 1420752 bytes
```
</details>

grzuy commented on examples/vision/cifar10.exs:
```diff
@@ -14,9 +14,10 @@ defmodule Cifar do
     bin
     |> Nx.from_binary(type)
-    |> Nx.reshape({elem(shape, 0), 3, 32, 32})
+    |> Nx.reshape({elem(shape, 0), 3, 32, 32}, names: [:n, :c, :h, :w])
```
```suggestion
    |> Nx.reshape(shape, names: [:n, :c, :h, :w])
```

I think shape is already this way?

nickgnd replied:
Hey @grzuy thanks for your comment!

Honestly I don't remember anymore, I should run the livebook to check it 🙈

nickgnd commented:
Hey all 👋 
I forgot about this PR opened long time ago, I see that there are some conflicts and most likely my changes are now outdated. Going to close it.

Cheers ✌️
